contracts/EvmGolf.sol

```diff
@@ -0,0 +1,71 @@
+//SPDX-License-Identifier: Unlicense
+pragma solidity ^0.8.0;
+
+contract EvmGolf {
+    mapping(string => address) private _levels; // level name -> level address
+    mapping(address => uint) private _authors; // author -> level address
+    mapping(address => address) private _solutions; // level -> solution
+    mapping(address => address) private _records; // level -> author
+    mapping(address => uint) private _victories; // author -> number of victories
+
+    function registerLevel(address newLevel) external returns (bool) {
+        const levelABI = [ "function name() external returns (string memory)"];
+        const level = new ethers.Contract(newLevel, levelABI, ethers.getDefaultProvider());
+        
+        // Verify that level name is not already registered
+        if (_levels[level.name()]) { return false; }
+        
+        // Stores the address of the level in a levels array
+        _levels[level.name()] = newLevel;
+
+        // Stores the address of the author of the level in an authors mapping
+        _authors[msg.sender] = newLevel;
+
+    }
+
+    function playLevel(address level, address solution) external {
+        const levelABI = [ 
+            "function submit(address candidate) external returns (bool)"
+        ];
+        const levelContract = new ethers.Contract(newLevel, levelABI, ethers.getDefaultProvider());
+
+        // Should this use await? Would submit happen synchronously?
+        const success = levelContract.submit(solution);
```

Comment:
@ajsantander are async functions a thing? I was wondering how I know if submit is async or sync and so if I should add an await or not.